File: algorithms/searching/binary-search/OrderAgnosticBinarySearch.py
def order_agnostic_binary_search(arr, target):
    if not arr:
        return -1
    start,end = 0,len(arr)-1
    ascending = arr[start]<arr[end]
    while(start<=end):
        mid = start+(end-start)//2
        if arr[mid] == target:
            return mid
        if ascending:
            if target < arr[mid]:
                end = mid-1
            else:
                start = mid+1
        else:
            if target>arr[mid]:
                end = mid-1
            else:
                start = mid+1
    return -1
def main():
    arr = [10, 8, 6, 4, 2, 0, -2, -4]
    target = -2
    result = order_agnostic_binary_search(arr, target)
    print(result)

if __name__ == "__main__":
    main()

File: algorithms/searching/binary-search/OrderAgnosticBinarySearch.rs
fn main() {
    let arr: Vec<i32> = vec![1, 2, 4, 5, 7, 2, 7];
    let target: i32 = 7;
    let result: i32 = order_agnostic_binary_search(&arr, target);
    println!("Target found at index: {}", result);
}

fn order_agnostic_binary_search(arr: &[i32], target: i32) -> i32 {
    if arr.len() == 0 {
        return -1;
    }
    let mut start: i32 = 0;
    let mut end: i32 = arr.len() as i32 - 1;
    let is_ascending: bool = arr[start as usize] < arr[end as usize];
    while start <= end {
        let mid: i32 = start + (end - start) / 2;
        if arr[mid as usize] == target {
            return mid;
        }
        if is_ascending {
            if target < arr[mid as usize] {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        } else {
            if target > arr[mid as usize] {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
    }
    -1
}
